### PR TITLE
NEWTAG-472 FACT/FACTSET/Output Models

### PR DESCRIPTION
**Pull request by thpr**
- Add FACT token (incl .CLEAR)
- Add FACTSET token (incl .CLEARALL)
- Add PREFACT
- Add PREFACTSET
- Add Ouptut Actors and Models (pcgen.output)
- Add Formats (as plugins) to support FACT/FACTSET
- Add FACT/FACTSET output to the UI info screen(s)

Internal changes:
- Split WeaponProfFacet per the TODO (to support more generic behavior)
- Some method renaming on facets to support custom code elimination

### CODE REVIEW

File: code/src/itest/pcgen/output/testsupport/AbstractOutputTestCase.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+import junit.framework.TestCase;
+import pcgen.cdom.enumeration.CharID;
+import pcgen.cdom.enumeration.DataSetID;
+import pcgen.output.OutputSupport;
+import pcgen.output.publish.OutputDB;
+import freemarker.template.Configuration;
+import freemarker.template.Template;
+import freemarker.template.TemplateException;
+
+public abstract class AbstractOutputTestCase extends TestCase
```
Comment (jdempsey-au) [inline code/src/itest/pcgen/output/testsupport/AbstractOutputTestCase.java]: Could junit 4 be used here?

File: code/src/itest/plugin/lsttokens/datacontrol/SelectableTokenIntegrationTest.java [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+import pcgen.persistence.lst.CampaignSourceEntry;
+import pcgen.rules.context.ConsolidatedListCommitStrategy;
+import pcgen.rules.context.LoadContext;
+import pcgen.rules.context.RuntimeLoadContext;
+import pcgen.rules.context.RuntimeReferenceContext;
+import pcgen.util.enumeration.Visibility;
+import plugin.format.StringManager;
+import plugin.lsttokens.testsupport.TokenRegistration;
+
+public class SelectableTokenIntegrationTest extends TestCase
```
Comment (jdempsey-au) [inline code/src/itest/plugin/lsttokens/datacontrol/SelectableTokenIntegrationTest.java]: Extending TestCase is not needed for Junit 4 test classes.

File: code/src/java/pcgen/gui2/facade/Gui2InfoFactory.java [hunk 3]
```diff
@@ -2128,4 +2159,77 @@ public String getMovement(RaceFacade race)
...
+	private void appendFacts(HtmlInfoBuilder infoText, CDOMObject cdo)
+	{
+		Class<? extends CDOMObject> cl = cdo.getClass();
+		LoadContext context = Globals.getContext();
+		Collection<FactDefinition> defs =
+				context.getReferenceContext().getConstructedCDOMObjects(
+					FactDefinition.class);
+		for (FactDefinition<?, ?> def : defs)
+		{
+			if (cl.equals(def.getUsableLocation()))
```
Comment (jdempsey-au) [inline code/src/java/pcgen/gui2/facade/Gui2InfoFactory.java]: Does this cope with global as the fact file type?

File: code/src/java/pcgen/output/actor/StringKeyActor.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+import pcgen.cdom.base.CDOMObject;
+import pcgen.cdom.enumeration.StringKey;
+import pcgen.output.base.OutputActor;
+import freemarker.template.ObjectWrapper;
+import freemarker.template.TemplateModel;
+import freemarker.template.TemplateModelException;
+
+/**
+ * A StringKeyActor is designed to process an interpolation and convert that
+ * into a TemplateModel representing the contents of the specific Integer being
```
Comment (jdempsey-au) [inline code/src/java/pcgen/output/actor/StringKeyActor.java]: Presumably this should be "the specfiic String" ?

File: code/src/java/pcgen/rules/context/LoadContext.java [hunk 5]
```diff
@@ -131,4 +131,6 @@ public CampaignSourceEntry getCampaignSourceEntry(Campaign source,
 	public void addWriteMessage(String string);
 
 	public int getWriteMessageCount();
+
+	public void loadLocalToken(Object token);
```
Comment (jdempsey-au) [inline code/src/java/pcgen/rules/context/LoadContext.java]: A description could be helpful here.

**Comment by jdempsey-au** [comment]
Just a few suggestions listed above, but this looks amazing and ready for merging. Nice testing and a real jump in the Freemarker usage and LST flexibility.

**Comment by thpr** [comment]
Thanks.

Will address this eve, was focused on the other issues that showed up in
476 this morning (solved 2 of 3)

TP

**Comment by thpr** [comment]
Addressed the non-Junit ones.  I've had bad luck in the past with Junit 4 in eclipse, so still have the Junit 3 habit.  Of course, that was about 5 eclipse releases ago, so will look at Junit 4 going forward (though it may deserve just a major refresh across everything, since the vast majority of what we have is still Junit 3)

**Pull request merged**